jobs: TestJobSchedulerDaemonProcessesJobs failed

[(jobs).TestJobSchedulerDaemonProcessesJobs failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2190017&tab=buildLog) on [master@3ea329fa4e3d52dcfb9fc92ecd5c9229b8942b97](https://github.com/cockroachdb/cockroach/commits/3ea329fa4e3d52dcfb9fc92ecd5c9229b8942b97):

```
=== RUN   TestJobSchedulerDaemonProcessesJobs
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestJobSchedulerDaemonProcessesJobs668446127
--- FAIL: TestJobSchedulerDaemonProcessesJobs (1.17s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestJobSchedulerDaemonProcessesJobs668446127
    test_log_scope.go:63: use -show-logs to present logs inline
    job_scheduler_test.go:379: 
        	Error Trace:	job_scheduler_test.go:379
        	Error:      	Not equal: 
        	            	expected: int(5)
        	            	actual  : int64(0)
        	Test:       	TestJobSchedulerDaemonProcessesJobs

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestJobSchedulerDaemonProcessesJobs PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestJobSchedulerDaemonProcessesJobs.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(jobs).TestJobSchedulerDaemonProcessesJobs failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2192812&tab=buildLog) on [master@b82ab8f3412dd2e1367c41bce15651908be5c9c1](https://github.com/cockroachdb/cockroach/commits/b82ab8f3412dd2e1367c41bce15651908be5c9c1):

```
=== RUN   TestJobSchedulerDaemonProcessesJobs
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestJobSchedulerDaemonProcessesJobs118094212
--- FAIL: TestJobSchedulerDaemonProcessesJobs (1.17s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestJobSchedulerDaemonProcessesJobs118094212
    test_log_scope.go:63: use -show-logs to present logs inline
    job_scheduler_test.go:379: 
        	Error Trace:	job_scheduler_test.go:379
        	Error:      	Not equal: 
        	            	expected: int(5)
        	            	actual  : int64(0)
        	Test:       	TestJobSchedulerDaemonProcessesJobs

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestJobSchedulerDaemonProcessesJobs PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestJobSchedulerDaemonProcessesJobs.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(jobs).TestJobSchedulerDaemonProcessesJobs failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2193699&tab=buildLog) on [master@6c3dfc4058eeba68ae588893ff732e0d6dd067ab](https://github.com/cockroachdb/cockroach/commits/6c3dfc4058eeba68ae588893ff732e0d6dd067ab):

```
=== RUN   TestJobSchedulerDaemonProcessesJobs
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestJobSchedulerDaemonProcessesJobs883290501
--- FAIL: TestJobSchedulerDaemonProcessesJobs (2.47s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestJobSchedulerDaemonProcessesJobs883290501
    test_log_scope.go:63: use -show-logs to present logs inline
    job_scheduler_test.go:379: 
        	Error Trace:	job_scheduler_test.go:379
        	Error:      	Not equal: 
        	            	expected: int(5)
        	            	actual  : int64(0)
        	Test:       	TestJobSchedulerDaemonProcessesJobs

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestJobSchedulerDaemonProcessesJobs PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestJobSchedulerDaemonProcessesJobs.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(jobs).TestJobSchedulerDaemonProcessesJobs failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2195070&tab=buildLog) on [master@d8eb600164fd978ab1eaf15cac3cf2cc36ad48a9](https://github.com/cockroachdb/cockroach/commits/d8eb600164fd978ab1eaf15cac3cf2cc36ad48a9):

```
=== RUN   TestJobSchedulerDaemonProcessesJobs
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestJobSchedulerDaemonProcessesJobs932214312
--- FAIL: TestJobSchedulerDaemonProcessesJobs (2.30s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestJobSchedulerDaemonProcessesJobs932214312
    test_log_scope.go:63: use -show-logs to present logs inline
    job_scheduler_test.go:379: 
        	Error Trace:	job_scheduler_test.go:379
        	Error:      	Not equal: 
        	            	expected: int(5)
        	            	actual  : int64(0)
        	Test:       	TestJobSchedulerDaemonProcessesJobs

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestJobSchedulerDaemonProcessesJobs PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestJobSchedulerDaemonProcessesJobs.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(jobs).TestJobSchedulerDaemonProcessesJobs failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2196686&tab=buildLog) on [master@ca8fc7284e90033148dae68b57dbef12c598b97b](https://github.com/cockroachdb/cockroach/commits/ca8fc7284e90033148dae68b57dbef12c598b97b):

```
=== RUN   TestJobSchedulerDaemonProcessesJobs
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestJobSchedulerDaemonProcessesJobs462778384
--- FAIL: TestJobSchedulerDaemonProcessesJobs (2.37s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestJobSchedulerDaemonProcessesJobs462778384
    test_log_scope.go:63: use -show-logs to present logs inline
    job_scheduler_test.go:379: 
        	Error Trace:	job_scheduler_test.go:379
        	Error:      	Not equal: 
        	            	expected: int(5)
        	            	actual  : int64(0)
        	Test:       	TestJobSchedulerDaemonProcessesJobs

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestJobSchedulerDaemonProcessesJobs PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestJobSchedulerDaemonProcessesJobs.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(jobs).TestJobSchedulerDaemonProcessesJobs failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2197448&tab=buildLog) on [master@697e455f453581cf450d0bbf2903a00c6d9ebffd](https://github.com/cockroachdb/cockroach/commits/697e455f453581cf450d0bbf2903a00c6d9ebffd):

```
=== RUN   TestJobSchedulerDaemonProcessesJobs
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestJobSchedulerDaemonProcessesJobs756711647
--- FAIL: TestJobSchedulerDaemonProcessesJobs (2.35s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestJobSchedulerDaemonProcessesJobs756711647
    test_log_scope.go:63: use -show-logs to present logs inline
    job_scheduler_test.go:379: 
        	Error Trace:	job_scheduler_test.go:379
        	Error:      	Not equal: 
        	            	expected: int(5)
        	            	actual  : int64(0)
        	Test:       	TestJobSchedulerDaemonProcessesJobs

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestJobSchedulerDaemonProcessesJobs PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestJobSchedulerDaemonProcessesJobs.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/jobs/job_scheduler.go
+++ b/pkg/jobs/job_scheduler.go
@@ -72,7 +72,7 @@ const allSchedules = 0
 // scheduled jobs that should be started.
 func getFindSchedulesStatement(env scheduledjobs.JobSchedulerEnv, maxSchedules int64) string {
 	limitClause := ""
-	if maxSchedules > 0 {
+	if maxSchedules != allSchedules {
 		limitClause = fmt.Sprintf("LIMIT %d", maxSchedules)
 	}
 
@@ -236,8 +236,10 @@ func (s *jobScheduler) executeSchedules(
 	defer stats.updateMetrics(&s.metrics)
 
 	findSchedulesStmt := getFindSchedulesStatement(s.env, maxSchedules)
-	rows, cols, err := s.InternalExecutor.QueryWithCols(ctx, "find-scheduled-jobs", nil,
-		sqlbase.InternalExecutorSessionDataOverride{User: security.RootUser},
+	rows, cols, err := s.InternalExecutor.QueryWithCols(
+		ctx, "find-scheduled-jobs",
+		txn,
+		sqlbase.InternalExecutorSessionDataOverride{User: security.NodeUser},
 		findSchedulesStmt)
 
 	if err != nil {
@@ -252,8 +254,20 @@ func (s *jobScheduler) executeSchedules(
 			continue
 		}
 
+		sp, err := txn.CreateSavepoint(ctx)
+		if err != nil {
+			return err
+		}
+
 		if err := s.processSchedule(ctx, schedule, numRunning, stats, txn); err != nil {
-			// We don't know if txn is good at this point, so bail out.
+			log.Errorf(ctx, "error processing schedule %d: %+v", schedule.ScheduleID(), err)
+
+			if err := txn.RollbackToSavepoint(ctx, sp); err != nil {
+				return errors.Wrapf(err, "failed to rollback savepoint for schedule %d", schedule.ScheduleID())
+			}
+		}
+
+		if err := txn.ReleaseSavepoint(ctx, sp); err != nil {
 			return err
 		}
 	}

--- a/pkg/jobs/job_scheduler_test.go
+++ b/pkg/jobs/job_scheduler_test.go
@@ -26,7 +26,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -42,8 +41,6 @@ import (
 func TestJobSchedulerReschedulesRunning(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-
-	skip.WithIssue(t, 52959)
 
 	h, cleanup := newTestHelper(t)
 	defer cleanup()
@@ -84,7 +81,10 @@ func TestJobSchedulerReschedulesRunning(t *testing.T) {
 			// The job should not run -- it should be rescheduled `recheckJobAfter` time in the
 			// future.
 			s := newJobScheduler(h.cfg, h.env, metric.NewRegistry())
-			require.NoError(t, s.executeSchedules(ctx, allSchedules, nil))
+			require.NoError(t,
+				h.cfg.DB.Txn(context.Background(), func(ctx context.Context, txn *kv.Txn) error {
+					return s.executeSchedules(ctx, allSchedules, txn)
+				}))
 
 			if wait == jobspb.ScheduleDetails_WAIT {
 				expectedRunTime = h.env.Now().Add(recheckRunningAfter)
@@ -138,7 +138,10 @@ func TestJobSchedulerExecutesAfterTerminal(t *testing.T) {
 
 			// Execute the job and verify it has the next run scheduled.
 			s := newJobScheduler(h.cfg, h.env, metric.NewRegistry())
-			require.NoError(t, s.executeSchedules(ctx, allSchedules, nil))
+			require.NoError(t,
+				h.cfg.DB.Txn(context.Background(), func(ctx context.Context, txn *kv.Txn) error {
+					return s.executeSchedules(ctx, allSchedules, txn)
+				}))
 
 			expectedRunTime = cronexpr.MustParse("@hourly").Next(h.env.Now())
 			loaded = h.loadSchedule(t, j.ScheduleID())
@@ -175,7 +178,10 @@ func TestJobSchedulerExecutesAndSchedulesNextRun(t *testing.T) {
 
 	// Execute the job and verify it has the next run scheduled.
 	s := newJobScheduler(h.cfg, h.env, metric.NewRegistry())
-	require.NoError(t, s.executeSchedules(ctx, allSchedules, nil))
+	require.NoError(t,
+		h.cfg.DB.Txn(context.Background(), func(ctx context.Context, txn *kv.Txn) error {
+			return s.executeSchedules(ctx, allSchedules, txn)
+		}))
 
 	expectedRunTime = cronexpr.MustParse("@hourly").Next(h.env.Now())
 	loaded = h.loadSchedule(t, j.ScheduleID())
@@ -429,6 +435,70 @@ func TestJobSchedulerDaemonHonorsMaxJobsLimit(t *testing.T) {
 		return nil
 	})
 	stopper.Stop(ctx)
+}
+
+// returnErrorExecutor counts the number of times it is
+// called, and always returns an error.
+type returnErrorExecutor struct {
+	numCalls int
+}
+
+func (e *returnErrorExecutor) ExecuteJob(
+	_ context.Context,
+	_ *scheduledjobs.JobExecutionConfig,
+	_ scheduledjobs.JobSchedulerEnv,
+	schedule *ScheduledJob,
+	_ *kv.Txn,
+) error {
+	e.numCalls++
+	return errors.Newf("error for schedule %d", schedule.ScheduleID())
+}
+
+func (e *returnErrorExecutor) NotifyJobTermination(
+	_ context.Context,
+	_ int64,
+	_ Status,
+	_ scheduledjobs.JobSchedulerEnv,
+	_ *ScheduledJob,
+	_ sqlutil.InternalExecutor,
+	_ *kv.Txn,
+) error {
+	return nil
+}
+
+func (e *returnErrorExecutor) Metrics() metric.Struct {
+	return nil
+}
+
+var _ ScheduledJobExecutor = &returnErrorExecutor{}
+
+func TestJobSchedulerToleratesBadSchedules(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	h, cleanup := newTestHelper(t)
+	defer cleanup()
+
+	ctx := context.Background()
+
+	const executorName = "return_error"
+	ex := &returnErrorExecutor{}
+	defer registerScopedScheduledJobExecutor(executorName, ex)()
+
+	// Create few one-off schedules.
+	const numJobs = 5
+	scheduleRunTime := h.env.Now().Add(time.Hour)
+	for i := 0; i < numJobs; i++ {
+		s := h.newScheduledJobForExecutor("schedule", executorName, nil)
+		s.SetNextRun(scheduleRunTime)
+		require.NoError(t, s.Create(ctx, h.cfg.InternalExecutor, nil))
+	}
+	h.env.SetTime(scheduleRunTime.Add(time.Second))
+	daemon := newJobScheduler(h.cfg, h.env, metric.NewRegistry())
+	require.NoError(t,
+		h.cfg.DB.Txn(context.Background(), func(ctx context.Context, txn *kv.Txn) error {
+			return daemon.executeSchedules(ctx, numJobs, txn)
+		}))
+	require.Equal(t, numJobs, ex.numCalls)
 }
 
 func TestJobSchedulerDaemonUsesSystemTables(t *testing.T) {

--- a/pkg/sql/delegate/show_jobs.go
+++ b/pkg/sql/delegate/show_jobs.go
@@ -20,6 +20,14 @@ import (
 )
 
 func (d *delegator) delegateShowJobs(n *tree.ShowJobs) (tree.Statement, error) {
+	if n.Schedules != nil {
+		// Limit the jobs displayed to the ones started by specified schedules.
+		return parse(fmt.Sprintf(`
+SHOW JOBS SELECT id FROM system.jobs WHERE created_by_type='%s' and created_by_id IN (%s)
+`, jobs.CreatedByScheduledJobs, n.Schedules.String()),
+		)
+	}
+
 	sqltelemetry.IncrementShowCounter(sqltelemetry.Jobs)
 
 	if n.Schedules != nil {

